readthedocs/oauth/models.py

```diff
@@ -84,6 +87,78 @@ class GitHubAppInstallation(TimeStampedModel):
     class Meta(TimeStampedModel.Meta):
         verbose_name = _("GitHub app installation")
 
+    @cached_property
+    def service(self):
+        """Return the service for this installation."""
+        from readthedocs.oauth.services import GitHubAppService
+
+        return GitHubAppService(self)
+
+    def delete(self, *args, **kwargs):
+        """Override delete method to remove orphaned organizations."""
+        self.delete_repositories()
+        return super().delete(*args, **kwargs)
+
+    def delete_repositories(self, repository_ids: list[int] | None = None):
+        """
+        Delete repositories linked to this installation.
+        When an installation is deleted, we delete all its remote repositories
+        and relations, users will need to manually link the projects to each repository again.
+        We also remove organizations that don't have any repositories after removing the repositories.
+        :param repository_ids: List of repository ids (remote ID) to delete.
+         If None, all repositories will be considered for deletion.
+        """
+        if repository_ids is not None and not repository_ids:
+            log.info("No repositories to delete")
+            return
+
+        remote_organizations = RemoteOrganization.objects.filter(
+            repositories__github_app_installation=self,
+            vcs_provider=GITHUB_APP,
+        )
+        remote_repositories = self.repositories.filter(vcs_provider=GITHUB_APP)
+        if repository_ids:
+            remote_organizations = remote_organizations.filter(
+                repositories__remote_id__in=repository_ids
+            )
+            remote_repositories = remote_repositories.filter(remote_id__in=repository_ids)
+
+        # Fetch all IDs before deleting the repositories, so we can filter the organizations later.
+        remote_organizations_ids = list(remote_organizations.values_list("id", flat=True))
+
+        count, deleted = remote_repositories.delete()
+        log.info(
+            "Deleted repositories projects",
+            count=count,
+            deleted=deleted,
+            installation_id=self.installation_id,
+        )
+
+        count, deleted = RemoteOrganization.objects.filter(
+            id__in=remote_organizations_ids,
+            repositories=None,
+        ).delete()
+        log.info(
+            "Deleted orphaned organizations",
+            count=count,
+            deleted=deleted,
+            installation_id=self.installation_id,
+        )
+
+    def delete_organization(self, organization_id: int):
+        """Delete an organization and all its repositories and relations from the database."""
+        count, deleted = RemoteOrganization.objects.filter(
+            remote_id=str(organization_id),
+            vcs_provider=GITHUB_APP,
+        ).delete()
+        log.info(
+            "Deleted organization",
```

Comment:
```suggestion
            "Deleted RemoteOrganizations",
```

Reply:
These log messages sounds scary. We should probably add more context where possible.

Reply:
I think something like `Deleted RemoteOrganizations because application was uninstalled from GitHub` or similar would make it more explicit.

Reply:
The reason for a deletion depends on the caller (repo deselected/app uninstalled/revoked, etc), I already log the installation/event/action in all webhooks https://github.com/readthedocs/readthedocs.org/blob/8b95371af062a88051b9675914e64a71631348e8/readthedocs/oauth/views.py#L65-L69